### ISSUE TITLE
compilation error with MinGW

### ISSUE THREAD

**Issue by ericLemanissier** [STEllAR-GROUP/hpx#1756]
It seems I can't compile HPX 0.9.10 under MinGW-w64 (i686-5.1.0-posix-dwarf). the error happens in hpx/util/coroutine/detail/default_context_impl.hpp : "No default_context_impl available for this system"

I can't use HPX_HAVE_GENERIC_CONTEXT_COROUTINES because Boost.Context is not compatible with MinGW, and there is no way to define HPX_HAVE_FIBER_BASED_COROUTINES as a cmake parameter when not using msvc (although if this symbol is set, compilation should just be ok.

**Comment by sithhell** [comment]
Am 21.09.2015 1:59 nachm. schrieb "ericLemanissier" <
notifications@github.com>:
>
> It seems I can't compile HPX 0.9.10 under MinGW-w64
(i686-5.1.0-posix-dwarf). the error happens in
hpx/util/coroutine/detail/default_context_impl.hpp : "No
default_context_impl available for this system"
>
> I can't use HPX_HAVE_GENERIC_CONTEXT_COROUTINES because Boost.Context is
not compatible with MinGW, and there is no way to define
HPX_HAVE_FIBER_BASED_COROUTINES as a cmake parameter when not using msvc
(although if this symbol is set, compilation should just be ok.

If compiling and running HPX on this platform with the mentioned define
works, we'd be happy do accept a patch for that!

**Comment by ericLemanissier** [comment]
Adding -DCMAKE_CXX_FLAGS="-DHPX_HAVE_FIBER_BASED_COROUTINES" to the cmake call removed this error, unfortunately I get quite a big number of other errors:
`conflicting declaration of snprintf with 'C'/'C++' linkage`
`include/boost-1_58/boost/smart_ptr/detail/spinlock_w32.hpp:52:7: error: redefinition of 'class boost::detail::spinlock'`
`include/boost-1_58/boost/smart_ptr/detail/spinlock_sync.hpp:30:7: error: previous definition of 'class boost::detail::spinlock'`

etc.

Should I understand that MinGW is not supported yet ? It mentioned in the documentation http://stellar-group.github.io/hpx/docs/html/hpx/manual/build_system/cmake_basics.html

**Comment by hkaiser** [comment]
@ericLemanissier: MinGW is not supported at this point. However, this is not because it does not work but just because nobody has invested the time to make it work. 

Would you be able/interested in contributing MinGW support for HPX?

**Comment by ericLemanissier** [comment]
I'd be glad to help if I can. I started some adaptations here : https://github.com/ericLemanissier/hpx/commits/mingw_support
I found a solution for spinlock, and _snprintf, but I'm now banging my head against the following:

```
hpx/traits/is_callable.hpp:290:12: error: ambiguous template instantiation for 'struct hpx::traits::detail::is_callable_impl<hpx::threads::thread_state_enum (hpx::lcos::detail::task_base<hpx::lcos::future<long long int> >::* const(const boost::intrusive_ptr<hpx::lcos::detail::task_base<hpx::lcos::future<long long int> > >&))(), void, void>'
     struct is_callable<F(A...)>
            ^
hpx/traits/is_callable.hpp:219:16: note: candidates are: template<class F, class C> struct hpx::traits::detail::is_callable_impl<F(C), typename hpx::util::always_void<decltype (*(declval<C>)().*(declval<F>)())>::type> [with F = hpx::threads::thread_state_enum (hpx::lcos::detail::task_base<hpx::lcos::future<long long int> >::* const)(); C = const boost::intrusive_ptr<hpx::lcos::detail::task_base<hpx::lcos::future<long long int> > >&]
         struct is_callable_impl<F(C)
                ^
hpx/traits/is_callable.hpp:249:16: note:                 template<class F, class C, class ... A> struct hpx::traits::detail::is_callable_impl<F(C, A ...), typename hpx::util::always_void<decltype (*(declval<C>)().*(declval<F>)()((declval<A>)()...))>::type> [with F = hpx::threads::thread_state_enum (hpx::lcos::detail::task_base<hpx::lcos::future<long long int> >::* const)(); C = const boost::intrusive_ptr<hpx::lcos::detail::task_base<hpx::lcos::future<long long int> > >&; A = {}]
         struct is_callable_impl<F(C, A...)
```
Any hints to solving this ? I'm compiling with -std=c++14

**Comment by ericLemanissier** [comment]
MinGW 4.9.2 returns the same error

**Comment by hkaiser** [comment]
@K-ballo Any insights? Will that be solved after merging #1749?

**Comment by K-ballo** [comment]
It looks similar to the Intel ambiguous specialization issue. My workaround for that one was simply disabling the C++11 implementation for Intel, since I had #1749 laying around and I expected it to solve the issue. If that doesn't solve the issue (it does for Intel), at least it will make the diagnostic more manageable.

**Comment by ericLemanissier** [comment]
Well, I cherry-picked my adaptations on top of branch result_of-invoke and I don't get the error any more, so I would say #1749 solves the issue on MinGW too !

**Comment by ericLemanissier** [comment]
I do not have any compilation error any more.
I can link libhpx.dll and libhpx_partitioned_vector.dll wihout problem, but i get errors for libhpx_parcel_coalescing.dll, libhpx_unordered.dll and libhpx_iostreams.dll:

```
CMakeFiles\parcel_coalescing_lib.dir/objects.a(coalescing_message_handler.cpp.obj):coalescing_message_handler.cpp:(.text$_ZNK3hpx7actions18typed_continuationIvE4saveERNS_13serialization14output_archiveEj[__ZNK3hpx7actions18typed_continuationIvE4saveERNS_13serialization14output_archiveEj]+0x1d):
 undefined reference to `void hpx::naming::id_type::save<hpx::serialization::output_archive>(hpx::serialization::output_archive&, unsigned int) const'
```
and the same for
```
void hpx::naming::id_type::load<hpx::serialization::input_archive>(hpx::serialization::input_archive&, unsigned int)
```
and
```
void hpx::naming::gid_type::load<hpx::serialization::input_archive>(hpx::serialization::input_archive&, unsigned int)
```

Have you any clue how to solve this ?

**Comment by sithhell** [comment]
On 09/24/2015 10:06 AM, ericLemanissier wrote:
> I do not have any compilation error any more.
> I can link libhpx.dll and libhpx_partitioned_vector.dll wihout problem, but i get errors for libhpx_parcel_coalescing.dll, libhpx_unordered.dll and libhpx_iostreams.dll:
> 
> ```
> CMakeFiles\parcel_coalescing_lib.dir/objects.a(coalescing_message_handler.cpp.obj):coalescing_message_handler.cpp:(.text$_ZNK3hpx7actions18typed_continuationIvE4saveERNS_13serialization14output_archiveEj[__ZNK3hpx7actions18typed_continuationIvE4saveERNS_13serialization14output_archiveEj]+0x1d):
>  undefined reference to `void hpx::naming::id_type::save<hpx::serialization::output_archive>(hpx::serialization::output_archive&, unsigned int) const'
> ```
> and the same for
> ```
> void hpx::naming::id_type::load<hpx::serialization::input_archive>(hpx::serialization::input_archive&, unsigned int)
> ```
> and
> ```
> void hpx::naming::gid_type::load<hpx::serialization::input_archive>(hpx::serialization::input_archive&, unsigned int)
> ```
> 
> Have you any clue how to solve this ?

This is strange. Those functions are defined in
src/runtime/naming/name.cpp. However, the unsigned int carries an
additional "const". Could you try to remove that and see what happens?

> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/STEllAR-GROUP/hpx/issues/1756#issuecomment-142847832
> 


-- 
Thomas Heller
Friedrich-Alexander-Universität Erlangen-Nürnberg
Department Informatik - Lehrstuhl Rechnerarchitektur
Martensstr. 3
91058 Erlangen
Tel.: 09131/85-27018
Fax:  09131/85-27912
Email: thomas.heller@cs.fau.de

**Comment by ericLemanissier** [comment]
I get the same linker error with the modifications in https://github.com/ericLemanissier/hpx/commit/cb2de9b0381af4536dde73005e4612d0667f47fc
Did I understand correctly what you asked me ?

It looks as if name.cpp.obj was not part of the object files passed to the linker. I can't find it in CMakeFiles\parcel_coalescing_lib.dir\linklibs.rsp nor CMakeFiles\parcel_coalescing_lib.dir\objects1.rsp whereas it is listed in CMakeFiles\hpx.dir\objects1.rsp

actually, CMakeFiles\parcel_coalescing_lib.dir\objects1.rsp contains only CMakeFiles/parcel_coalescing_lib.dir/plugins/parcel/coalescing/coalescing_message_handler.cpp.obj
Is this normal behavior ?

**Comment by hkaiser** [comment]
> It looks as if name.cpp.obj was not part of the object files passed to the linker.

Yes, that's expected, it is linked into libhpx.

**Comment by sithhell** [comment]
Am 24.09.2015 5:13 nachm. schrieb "ericLemanissier" <
notifications@github.com>:
>
> I get the same linker error with the modifications in
ericLemanissier@cb2de9b
> Did I understand correctly what you asked me ?

Yes.

>
> It looks as if name.cpp.obj was not part of the object files passed to
the linker. I can't find it in
CMakeFiles\parcel_coalescing_lib.dir\linklibs.rsp nor
CMakeFiles\parcel_coalescing_lib.dir\objects1.rsp whereas it is listed in
CMakeFiles\hpx.dir\objects1.rsp
>
> actually, CMakeFiles\parcel_coalescing_lib.dir\objects1.rsp contains only
CMakeFiles/parcel_coalescing_lib.dir/plugins/parcel/coalescing/coalescing_message_handler.cpp.obj
> Is this normal behavior ?

name.cpp.obj is contained in the main hpx dynamic library only and should
be visible. Might be a bug with the elf visibility (set in cmake). Is this
turned on when compiling with mingw?

**Comment by hkaiser** [comment]
Actually, the functions in question are currently not explicitly exported (see https://github.com/STEllAR-GROUP/hpx/blob/master/src/runtime/naming/name.cpp#L562). I'll fix this asap.

**Comment by sithhell** [comment]
Am 24.09.2015 5:25 nachm. schrieb "Hartmut Kaiser" <notifications@github.com
>:
>
> Actually, the functions in question are currently not explicitly exported
(see
https://github.com/STEllAR-GROUP/hpx/blob/master/src/runtime/naming/name.cpp#L562).
I'll fix this asap.

id_type and gid_type are exported. It seems to work on all other platforms.

**Comment by ericLemanissier** [comment]
I am actually linking with the flag -fvisibility=hidden, that could be the explanation

**Comment by hkaiser** [comment]
> I am actually linking with the flag -fvisibility=hidden, that could be the explanation.

Could you please verify whether this is causing the problems?

**Comment by ericLemanissier** [comment]
https://github.com/STEllAR-GROUP/hpx/commit/abeef50bd9e0498550b8c95b8d6bc417ad0cd68a solved the problem, at least with -DHPX_WITH_HIDDEN_VISIBILITY=OFF

**Comment by ericLemanissier** [comment]
Very quickly, I am now hitting undefined references in the examples for data_actions
```
../../lib/libhpx_init.a(main.cpp.obj):main.cpp:(.text$_ZN3hpx4initEiPPcRKSt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS8_EENS_12runtime_modeE[__ZN3hpx4initEiPPcRKSt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS8_EENS_12runtime_modeE]+0xa7): undefined reference to `_imp___ZN3hpx4util14set_hpx_prefixEPKc'
../../lib/libhpx_init.a(main.cpp.obj):main.cpp:(.text$_ZN3hpx4initEiPPcRKSt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS8_EENS_12runtime_modeE[__ZN3hpx4initEiPPcRKSt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS8_EENS_12runtime_modeE]+0xe9): undefined reference to `_imp___ZN3hpx6detail12run_or_startERKNS_4util8functionIFiRN5boost15program_options13variables_mapEEvvEERKNS4_19options_descriptionEiPPcRKSt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaISM_EERKNS2_IFvvEvvEESU_NS_12runtime_modeEb'
../../lib/libhpx_init.a(hpx_main.cpp.obj):hpx_main.cpp:(.text+0x6ae): undefined reference to `_imp___ZN3hpx11get_runtimeEv'
../../lib/libhpx_init.a(hpx_main.cpp.obj):hpx_main.cpp:(.text+0x716): undefined reference to `_imp___ZNK3hpx4util7section9get_entryERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES9_'
../../lib/libhpx_init.a(hpx_main_argc_argv.cpp.obj):hpx_main_argc_argv.cpp:(.text+0x6d0): undefined reference to `_imp___ZN3hpx8finalizeEddRNS_10error_codeE'
```

**Comment by ericLemanissier** [comment]
ok, it was just a matter of switching the order of libhpx and libhpx_init in the linker arguments. It now compiles flawlessly on MinGW, but I could not check that it still works on other compiler. I just made the pull request https://github.com/STEllAR-GROUP/hpx/pull/1773

**Comment by hkaiser** [comment]
This was fixed by merging #1773
